### PR TITLE
Add "Freeze Frame" mod

### PR DESCRIPTION
**Pull request by mk56-spn**
As discussed in #19788, this new mod ( which ive called Freeze Frame since I think it fits its function very well, which is to sort of shot a snapshot of the playfield at the beginning of each control point ( or half, double, quadruple etc.. depending on the mod settings )) provides an interesting play experience where note order is now something you have to deduce from either memory, note numbers, pattern recognition, or any other form of "static" indicator.

With that in mind one may ask why I opted not to remove approach circles. the  reason for that is simple. although they no longer denote the time left till the next note and as such aren't as reliable as they usually are they still give a vague indication of what comes first based on size, the speed at which the approach circle closes in, and help with perfectly stacked notes.

**Notable mod combinations:** 

+HD : I expect this mod will mostly be played with Hidden as to reduce the amount of notes on screen and since, for those who dont want them, it offers the option to remove the approach circles, so overall a decent companion to this

+HDFL , having limited time to pass over the notes  in whatever breaks the song may offer will allow the player to have a  bit of time to see whats coming for the next beat length.

**Code Quality related stuff** 

I ended up opting for an enum drop down because sliders do not allow specific values to be the only ones the user can select. which is crucial here to avoid unnecessary beat length combinations that are out of sync with the song.

_a little fun test with FFTCAD :P :_

https://user-images.githubusercontent.com/74463310/190710729-25abdc4d-43e6-4a6b-b044-17931d6c797b.mov

### CODE REVIEW

**Comment by mk56-spn** [comment]
Sorry about the less than compact commit history 😅

**Comment by frenzibyte** [comment]
> Sorry about the less than compact commit history 😅

I would highly suggest rewriting commit history before submitting PR, or at the very least squashing all changes to one commit.

**Comment by mk56-spn** [comment]
> > Sorry about the less than compact commit history 😅
> 
> I would highly suggest rewriting commit history before submitting PR, or at the very least squashing all changes to one commit. PR's been opened up now so there's probably no need to do it.

Yep usually I'd squash but this branch is pretty old and I had pushed quite a few to remote and didn't want to mess anything up with rebases which I'm not too comfortable with. If it's too messy I'll close the PR and just make a clean branch

**Comment by loma-baldsson** [comment]
Not sure where I should put this, but I found that reverse arrows are sometimes drawn on top of the hitcircle on kick sliders.

https://user-images.githubusercontent.com/102317848/190877025-08e1304b-320d-47a8-8c81-59c864daf122.mp4

**Comment by mk56-spn** [comment]
> Not sure where I should put this, but I found that reverse arrows are sometimes drawn on top of the hitcircle on kick sliders.

hmm, well the mod doesnt affect depth explicitly so i assume this has something to do with the draw order being down to which is drawn first.

**for maintainers**
 should this be fixed at a higher level if this is indeed the case or should i try something hacky here?

**Review by ggliv — CHANGES_REQUESTED** [review]
didn't really look in to the actual mod logic but here are some things I saw at a glance. weird mod, I like it.

File: osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs
Comment (mk56-spn) [reply osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs:23]: Sounds ok ye

File: osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs
Comment (Walavouchey) [reply osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs]: they're called "hit circles" (refer to the wiki)

**Comment by peppy** [comment]
The default settings have no noticeable effect on most beatmaps I tested. Oversight?

Rather than adding settings for new mods in the initial PR, how about making the mod usable *without* changing settings first? That should be the default mode of operation. A user should *not* have to change settings to make the effect of a mod visible, or to make it usable with a specific difficulty.

Please remove the setting and make the default experience worthwhile.

**Comment by mk56-spn** [comment]
> The default settings have no noticeable effect on most beatmaps I tested. Oversight?
> 
> Rather than adding settings for new mods in the initial PR, how about making the mod usable _without_ changing settings first? That should be the default mode of operation. A user should _not_ have to change settings to make the effect of a mod visible, or to make it usable with a specific difficulty.
> 
> Please remove the setting and make the default experience worthwhile.

it was specifically chosen to be very soft on the default setting, it has a clear effect on most maps at that rang,e notably and 1/2 jump and all streams are affected by it , 

furthermore it has to be a setting because of songs like big black where one version is timed as 360 bpm and another is timed as 180 bpm, i can not create an enjoyable experience for both without a setting, 

as an addition to that, some songs like " https://osu.ppy.sh/beatmapsets/332532#osu/736217" map  what most would consider 1/2 rythm jumps to 1/1, again i can not account for that without a setting i wasnt just adding for the sake of it. 

now im not going to sit here and try and earn myself a ban from the repo by arguing , i just want it to be clear that i didnt select what i did willy nilly, 


i can make the default setting 2 instead of 1 but that would feel too harsh for most maps people play ( its a pretty hard mod so i expect most people playing it would be atleast in the 5 star+ skill range)

**Comment by peppy** [comment]
Have you considered making it operated based on object count rather than time interval?

**Comment by mk56-spn** [comment]
> Have you considered making it operated based on object count rather than time interval?

I cant see this being the way forward assuming you mean going purely only off object count without any beat sync, with object count it would kill any intuitive ability to read with the mod, hitobjects no longer have a routine fixed anchor you can sync to if you you know how spaced the beats are in time, let me illustrate:

indicator

up top we have in red the line of time signature based "freeze" moments, its regular as one would assume, or as regular as the maps bpm allows it to be

in orange is the mod if object count was used ( here i used 3 objects, although theres more issues such as whether slider ends, repeats etc should count...) as you can see where the next freeze lands is pretty much up in the air, (as far as humans are concerned anyways) in anything bar a regular stream or group of jumps, furthermore, if done by object count a section with slow jumps would leave you in the same freeze for potentially multiple seconds whilst then getting flickering for the whole of every stream in the map.

because of this i dont see object count being viable, unless you meant doing it by note density, but at that point i think the formula becomes so convoluted that itll make the mod something people dont really want to play

in the end its up to you ofcourse, but id heavily discourage this direction based on my 5~ hours of playtesting

**Comment by Feodor0090** [comment]
I'm proposing to base this on combo chains. Once last note in a combo is pressed/missed, show all notes/sliders from next combo.

**Comment by mk56-spn** [comment]
> I'm proposing to base this on combo chains. Once last note in a combo is pressed/missed, show all notes/sliders from next combo.

Now this is something I can get onboard with with a small condition. Not on whether it's missed or hit. Simply the Time preempt before the new combo. That way it's actually playable

**Comment by mk56-spn** [comment]
https://user-images.githubusercontent.com/74463310/191334492-48df8b2b-9cdc-403d-8842-29a2a658802c.mov

Changed to work with New Combos as suggested above. ill push when / if ppy agrees with this being an ok direction to take the mod in to avoid more unnecessary commits

**Comment by Feodor0090** [comment]
How spinners are handled? In triangles skin, i would expect them to appear in "contracted" state with all other hitobjects (i mean visual state as on screenshot below). 
изображение

Once it's time to start the spinner, it should play expanding animation as usual. But i don't know how to do this with legacy skins - spinners there have no preemptive animation, they just fade in from nowhere once it's time to. Argon spinners which are currently not implemented are also in question.

**Comment by mk56-spn** [comment]
> > Simply the Time preempt before the new combo
> 
> Yeah, it's fair - player won't face notes with too high AR. But if we are basing on combo, the mod creates an experience "hit everything in this combo to see the next one". If "next one" is appearing before you deal with "this", things start messing up. I still insist that making next combo appearing not earlier than previous one ends will improve player's experience.

i simply cant see myself agreeing to having notes pop in on something as arbitrary as the last judgement that will vary considerably depending on player input, and making reading flat out impossible on lots of things like streams without memorising, for the sake of a pop in effect. i will change this only if forced to, i dont consider it in the best interest of the majority of players 

as for the spinner im experimenting with i

**Comment by HiddenNode** [comment]
How about keeping the mod as is, and choosing the optimal "default" setting for each map based on something like the average note density of the map, that would solve both the problem mentionned by ppy, and the problem you mentionned about certain songs being mapped at different BPMs.

**Comment by mk56-spn** [comment]
That's also an option .I'm open to suggestions. Just not sure how we can reach a consensus..

**Comment by Doksha** [comment]
There should also be an option to freeze approach circles (maybe even slider balls) this was my initial intention with the mod, to have it switch to the next time frame, and that frame being the only information you get (as well as cursor location). Approach circles and slider balls moving just makes the mod feel weird.

**Comment by mk56-spn** [comment]
> There should also be an option to freeze approach circles (maybe even slider balls) this was my initial intention with the mod, to have it switch to the next time frame, and that frame being the only information you get (as well as cursor location). Approach circles and slider balls moving just makes the mod feel weird. This is more of an "Explosion" than a "Freeze Frame" or "Blink" mod.

whats the point in frozen approach circles. seems like it would add nothing to the gp over hidden (AO) , other than a mildly amusing visual effect that clutters the screen

slider balls moving i dont have a strong opinion on

as for the naming scheme, many mods dont exactly match up to the dictionary definition with what they do. 

still i understand it was your idea so i get why youd feel strongly about the direction it was taken. to anyone else surveying this, a third opinion would be greatly appreciated

**Comment by Doxxz** [comment]
> whats the point in frozen approach circles
When included together with frozen slider balls, it gives a better illusion of a "frozen frame".

**Comment by Doksha** [comment]
I want it to have a feel somewhat like the chimp test, you have all the circles for a small section pop on screen, and you have to hit them in the right order and to the rhythm (the circles should also fade out like hidden, but having the hit circles is basically necessary). approach circles moving would kinda defeat the purpose as you could just play off the hit circles.

**Comment by mk56-spn** [comment]
> I want it to have a feel somewhat like the chimp test, you have all the circles for a small section pop on screen, and you have to hit them in the right order and to the rhythm (the circles should also fade out like hidden, but having the hit circles is basically necessary). approach circles moving would kinda defeat the purpose as you could just play off the hit circles.

as ive said before , for circle fading you can stick hidden on the mod. doubling up on existing functionality for tiny changes doesnt feel right

approach circles give away much less info than they usually do due to the ar basically being variable ( new AR = base AR + dist in time from Last new combo )

and as ive also already mentioned they can be removed with the approach fade only setting. 

As for freezing them im not in favour

**Comment by mk56-spn** [comment]
https://youtu.be/7xUnl8lrpCs?t=54

fwiw ive done some more playtesting and the recent readdition of notes  being dimmed until you are within the hit window allows for some cheesing , since you can use it to know which notes you have to hit next

https://youtu.be/7xUnl8lrpCs?t=54 timestamp for those not in the know

**Comment by peppy** [comment]
I feel like setting the divisor to 4 gives much more interesting gameplay. To the point I think it should be the default value for this mod. Would you be against that?

**Comment by mk56-spn** [comment]
> I feel like setting the divisor to 4 gives much more interesting gameplay. To the point I think it should be the default value for this mod. Would you be against that?

Id be ok with that, but most people seem to  prefer the  "change every new  combo " option that was suggested above by feodor, incase you missed it. (https://github.com/ppy/osu/pull/20345#issuecomment-1252743953) but if thats not something you can level with then yes id be ok with changing the default to 4

**Comment by peppy** [comment]
Change every combo sounds better, if that comes alongside removing the divisor option completely.

File: osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs:52]: This seems like an error and may cause other nested objects to not be processed:

```suggestion
                            break;
```

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs
Comment (mk56-spn) [reply osu.Game.Rulesets.Osu/Mods/OsuModFreezeFrame.cs]: Fixed.

**Comment by Feodor0090** [comment]
It does not work with TR in it's current state. I'm expecting that all circles will start moving from screen sides at once. I'm guessing, TR just overwrites FR changes to hitobjects?

**Comment by mk56-spn** [comment]
> It does not work with TR in it's current state. I'm expecting that all circles will start moving from screen sides at once. I guess, TR just overwrites FR changes to hitobjects?

good catch, i dont think its overwriting, i think its just mods dont know what other mods have done to settings, so it probably pulls the default approach rate. not sure what can be done about this , this is going to become an increasingly common issue with each mod added, that being that the desired interactions isnt achieved because they seem not to know about eachother

**Comment by Feodor0090** [comment]
About notes dimming - it definetly should be disabled somehow. If you try to play TPFR, you will see that it's unreasonable hard because you can't read the numbers on circles. At all. I even can't see which circle is the first, because it fades in (and the number becames visible) too late.
Otherwise, on normal maps which have less crazy patterns and numbers are not so important dimming makes the game easier because it directly says you what to click next (already stated above).

Also, the mod works as it should only with long combos. If we have 12 12 12 123, there is no difference from normal gameplay. What about making a threshold about combo length and group 2-3 combos if they are too short? 5 notes maybe? So, 1212 and 12123 will appear at once.

**Comment by peppy** [comment]
I think it's fine for it to "not work" with short combos, since that's a very edge case in the grand scheme of things.

**Comment by peppy** [comment]
So I think this is in a pretty good state now, but having the approach rate variable feels a bit weird. I'm okay with this if others are on board with it, but would at least want @smoogipoo's opinion on the matter.

**Comment by smoogipoo** [comment]
If the video shown in https://github.com/ppy/osu/pull/20345#issuecomment-1252743953 is a representation of the variable approach rate in its current state, I did think that was weird at first. But I'm fine with it if it works better with it than without it.

**Comment by mk56-spn** [comment]
> If the video shown in https://github.com/ppy/osu/pull/20345#issuecomment-1252743953 is a representation of the variable approach rate in its current state, I did think that was weird at first. But I'm fine with it if it works better with it than without it.
> 
> Edit: It does play weird imo.

I've spent about 20 hours playing it now and whilst it is somewhat weird it's still more helpful for learning the mod and playing it than no approach circle ( which players can access with hiddens approach circle only option) and there's no other way to avoid it because the mod is dilating the amount of space between a hitcircle begining and the time it's hit to align a bunch of starttimes so they suddenly appear. Variable AR is simply put a consequence of the concept of the mod

**Comment by peppy** [comment]
Just make the approach circles appear as per usual but only when you get close in time to the object? What's the worry? It will become too easy? I don't think that's a concern.

**Comment by mk56-spn** [comment]
> Just make the approach circles appear as per usual but only when you get close in time to the object? What's the worry? It will become too easy? I don't think that's a concern.

That will definitely make it far too easy I guarantee you. Like just NM but with a bit more visual cluster easy

**Comment by peppy** [comment]
Well all I can say at this point is I don't think it will get merged with the variable AR. It just feels off-putting in a mod that otherwise is amazing.

**Comment by mk56-spn** [comment]
> Well all I can say at this point is I don't think it will get merged with the variable AR. It just feels off-putting in a mod that otherwise is amazing.

Your call to make obviously. I personally think it detracts from the playing experience and waters down the mod and won't pretend I agree it's the way forward but if this is final then I guess it is what it is

**Comment by mk56-spn** [comment]
https://user-images.githubusercontent.com/74463310/196158923-9755c66c-b34c-4bc6-a712-146018c63a88.mov

@peppy is this more akin to how you imagined it being? the approach circles appear as they would "normally"

ive alread voiced my concerns aplenty so i just want a confirmation that this is indeed whats desired before i clean up the code and fix a few hacky things i did to make this "demo" of how it would look

**Comment by peppy** [comment]
Yes, that looks fine.

**Comment by Doksha** [comment]
How difficult would it be to make the original implementation (everything frozen, with AR enough for you to see all notes in a block, with blocks being 1 measure in length, while the block rendering updates every half measure)
I might just take the time to do it myself as I am (personally) kinda disapointed with how this turned out relative to my initial idea.

**Comment by mk56-spn** [comment]
> How difficult would it be to make the original implementation (everything frozen, with AR enough for you to see all notes in a block, with blocks being 1 measure in length, while the block rendering updates every half measure) I might just take the time to do it myself as I am (personally) kinda disapointed with how this turned out relative to my initial idea.

again its not about difficulty, its about compromising to get it in, the reason i havent done it is because it doesnt seem to be wanted.

**Comment by peppy** [comment]
I've been hesitant to merge this one because I agree to an extent that the "soul" of this mod was lost a bit with the recent changes. I've pushed a potential path forward, which still has the approach circles fade in with the frozen frame, but also does not change this AR (instead making the approach circles start larger than they normally would).

Interested in thoughts on this path. I think I prefer it?

**Comment by mk56-spn** [comment]
> I've been hesitant to merge this one because I agree to an extent that the "soul" of this mod was lost a bit with the recent changes. I've pushed a potential path forward, which still has the approach circles fade in with the frozen frame, but also does not change this AR (instead making the approach circles start larger than they normally would).
> 
> Interested in thoughts on this path. I think I prefer it?

seems workable, blurry approach circles aside since those arent explicitly the fault of the change, im happy with this being the  way it gets in, however i wonder if theres someway we can have hidden and approach different play well with the mod, since on a conceptual level they shouldnt be incompatible, especially since this mod with the hidden (approach circle hide only) on is something i foresee many people wanting to play. i tried to make it work  but i cant see a way to have the transforms not break eachother

**Comment by peppy** [comment]
The beatmap's approach rate should still be factored in, if that's what you mean?

**Comment by mk56-spn** [comment]
> The beatmap's approach rate should still be factored in, if that's what you mean?

not sure, what i mean is that since we have been clearing out all the scale transforms to apply new ones it means it cant work with approach different (such as still keeping the accelerate and deccelerate effects) , and i was wondering if their was some way to make those transforms "additive" to the ones being done here, minor concern for sure

**Comment by peppy** [comment]
Ah, sorry, I misread the "different" word as "rate" and was very confused as a result.

Yes, I think this could be modified to work with approach different, but probably fine to do that as a follow-up effort?

**Comment by mk56-spn** [comment]
> Ah, sorry, I misread the "different" word as "rate" and was very confused as a result.
> 
> Yes, I think this could be modified to work with approach different, but probably fine to do that as a follow-up effort?

Very much agree on leaving this for a follow up effort. Been quite the tricky mod already 😅

**Comment by peppy** [comment]
Let's get this in for testing in the next release. I think it's feeling pretty good.

**Review by peppy — APPROVED** [review]

**Pull request merged**
